src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+
+
+@numba.extending.box(IndexedOptionType)
+def IndexedOptionType_box(typ, val, c):
+    # get PyObject of the Indexed class
+    IndexedOption_obj = c.pyapi.unserialize(c.pyapi.serialize_object(IndexedOption))
+
+    builder = numba.core.cgutils.create_struct_proxy(typ)(
+        c.context, c.builder, value=val
+    )
+    index_obj = c.pyapi.from_native_value(typ.index, builder.index, c.env_manager)
+    content_obj = c.pyapi.from_native_value(typ.content, builder.content, c.env_manager)
+
+    out = c.pyapi.call_function_objargs(
+        IndexedOption_obj,
+        (
+            index_obj,
+            content_obj,
+        ),
+    )
+
+    # decref PyObjects
+    c.pyapi.decref(IndexedOption_obj)
+
+    c.pyapi.decref(index_obj)
+    c.pyapi.decref(content_obj)
+
+    return out
+
+
+@numba.extending.overload_method(IndexedOptionType, "_length_get", inline="always")
+def IndexedOption_length(builder):
+    def getter(builder):
+        return builder._index._length_pos[0]
+
+    return getter
+
+
+@numba.extending.overload_method(IndexedOptionType, "_index", inline="always")
+def IndexedOption_index(builder):
+    def getter(builder):
+        return builder._index
+
+    return getter
+
+
+@numba.extending.overload_method(IndexedOptionType, "append_valid")
+def IndexedOption_append_valid(builder):
+    if isinstance(builder, IndexedOptionType):
+
+        def append_valid(builder):
+            builder._index.append(len(builder._content))
+            return builder._content
+
+        return append_valid
+
+
+@numba.extending.overload_method(IndexedOptionType, "extend_valid")
+def IndexedOption_extend_valid(builder, size):
+    def extend_valid(builder, size):
+        start = len(builder._content)
+        stop = start + size
+        builder._index.extend(list(range(start, stop)))
+        return builder._content
+
+    return extend_valid
+
+
+@numba.extending.overload_method(IndexedOptionType, "append_invalid")
+def IndexedOption_append_invalid(builder):
+    if isinstance(builder, IndexedOptionType):
+
+        def append_invalid(builder):
+            builder._index.append(-1)
+
+        return append_invalid
+
+
+@numba.extending.overload_method(IndexedOptionType, "extend_invalid")
+def IndexedOption_extend_invalid(builder, size):
+    def extend_invalid(builder, size):
+        builder._index.extend([-1] * size)
+
+    return extend_invalid
+
+
+########## ByteMasked #########################################################
+
+
+@final
+class ByteMasked(LayoutBuilder):
+    def __init__(
+        self,
+        dtype,  # mask must be "bool"
+        content,
+        *,
+        valid_when=True,
+        parameters=None,
+        initial=1024,
+        resize=8.0,
+    ):
+        self._mask = GrowableBuffer(dtype=dtype, initial=initial, resize=resize)
+        self._content = content
+        self._valid_when = valid_when
+        self._parameters = parameters
```

Comment:
This is an interface that users will need to use, so we can leave out the `dtype` that they don't get control over.

```suggestion
@final
class ByteMasked(LayoutBuilder):
    def __init__(
        self,
        content,
        *,
        valid_when=True,
        parameters=None,
        initial=1024,
        resize=8.0,
    ):
        self._mask = GrowableBuffer(dtype=numpy.dtype(numpy.bool_), initial=initial, resize=resize)
        self._content = content
        self._valid_when = valid_when
        self._parameters = parameters
```